server/sonar-server/src/main/java/org/sonar/server/user/ws/CurrentAction.java

```diff
@@ -126,33 +136,74 @@ private CurrentWsResponse toWsResponse(DbSession dbSession, String userLogin) {
-    String homepageType = user.getHomepageType();
-    if (homepageType == null) {
+    if (noHomepageSet(user)) {
       return defaultHomepage();
     }
+
+    return doBuildHomepage(dbSession, user).orElse(defaultHomepage());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> doBuildHomepage(DbSession dbSession, UserDto user) {
+
+    if (PROJECT.toString().equals(user.getHomepageType())) {
+      return projectHomepage(dbSession, user);
+    }
+
+    if (APPLICATION.toString().equals(user.getHomepageType()) || PORTFOLIO.toString().equals(user.getHomepageType())) {
+      return applicationAndPortfolioHomepage(dbSession, user);
+    }
+
+    if (ORGANIZATION.toString().equals(user.getHomepageType())) {
+      return organizationHomepage(dbSession, user);
+    }
+
+    return of(CurrentWsResponse.Homepage.newBuilder()
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .build());
+  }
+
+  private Optional<CurrentWsResponse.Homepage> projectHomepage(DbSession dbSession, UserDto user) {
+    Optional<ComponentDto> projectOptional = ofNullable(dbClient.componentDao().selectByUuid(dbSession, of(user.getHomepageParameter()).orElse(EMPTY)).orNull());
+    if (!projectOptional.isPresent()) {
+      cleanUserHomepageInDb(dbSession, user);
+      return empty();
+    }
+
     CurrentWsResponse.Homepage.Builder homepage = CurrentWsResponse.Homepage.newBuilder()
-      .setType(CurrentWsResponse.HomepageType.valueOf(homepageType));
-    setHomepageParameter(dbSession, homepageType, user.getHomepageParameter(), homepage);
-    return homepage.build();
+      .setType(CurrentWsResponse.HomepageType.valueOf(user.getHomepageType()))
+      .setComponent(projectOptional.get().getKey());
+    setNullable(projectOptional.get().getBranch(), homepage::setBranch);
+    return of(homepage.build());
   }
 
-  private void setHomepageParameter(DbSession dbSession, String homepageType, @Nullable String homepageParameter, CurrentWsResponse.Homepage.Builder homepage) {
-    if (PROJECT.toString().equals(homepageType)) {
-      checkState(homepageParameter != null, "Homepage parameter should not be null");
-      ComponentDto component = dbClient.componentDao().selectByUuid(dbSession, homepageParameter)
-        .or(() -> {
-          throw new IllegalStateException(format("Unknown component '%s' for homepageParameter", homepageParameter));
-        });
-      homepage.setComponent(component.getKey());
-      setNullable(component.getBranch(), homepage::setBranch);
-      return;
+  private Optional<CurrentWsResponse.Homepage> applicationAndPortfolioHomepage(DbSession dbSession, UserDto user) {
+    Optional<ComponentDto> componentOptional = ofNullable(dbClient.componentDao().selectByUuid(dbSession, of(user.getHomepageParameter()).orElse(EMPTY)).orNull());
+    if (!componentOptional.isPresent() || !pluginRepository.hasPlugin(GOVERNANCE_PLUGIN_KEY)) {
```

Comment:
remark: it'd be nice to have naming convention about WS field, properties etc. It's not related to this variable in particular, as we don't have any convention I'm aware of